EIPS/eip-7792.md

```diff
@@ -0,0 +1,126 @@
+    block: BlockMeta
+    transaction_root: Root
+
+class Log(Container):
+    address: ExecutionAddress
+    topics: List[Bytes32, MAX_TOPICS_PER_LOG]
+    data: ByteList[MAX_LOG_DATA_SIZE]
+
+class LogEntry(Container):
+    meta: LogMeta
+    log: Log
+```
+
+The `hash_tree_root(LogEntry)` commitments are subsequently tracked as part of `LOG_CONTRACT_ADDRESS`.
+
+```python
+def accumulate_log(evm: Evm, entry_root: Bytes32, key: Bytes32):
+    root = hashlib.sha256()
+    root.update(entry_root)
+    root.update(sload(evm.env.state, LOG_CONTRACT_ADDRESS, key))
+    sstore(evm.env.state, LOG_CONTRACT_ADDRESS, key, root.digest())
+
+def track_log(evm: Evm, entry: LogEntry) -> None:
+    entry_root = entry.hash_tree_root()
+
+    # Allow verification via `address` filter
+    key = keccak256(abi.encode(entry.log.address, LOG_ADDRESS_STORAGE_SLOT))
+    accumulate_log(evm, entry_root, key)
+
+    for topic in entry.log.topics:
+        # Allow verification via `topics` filter
+        key = keccak256(abi.encode(topic, LOG_TOPICS_STORAGE_SLOT))
+        accumulate_log(evm, entry_root, key)
+
+        # Allow verification via combined `address` + `topics` filter
+        key = keccak256(abi.encode(entry.log.address, topic))
+        key = keccak256(abi.encode(key, LOG_ADDRESS_TOPICS_STORAGE_SLOT))
+        accumulate_log(evm, entry_root, key)
+```
+
+### JSON-RPC API
+
+The `eth_getLogs` response format is extended to include:
+
+- `blockTimestamp`: `QUANTITY` - The timestamp field of the block referred to by `blockHash`
+
+### Verification
+
+For `eth_getLogs(address, topics, fromBlock, toBlock)`, the response data can be verified for correctness and completion by obtaining:
+
+1. `fromBlock` and `toBlock` block headers (validated against their known hashes)
+2. `fromBlock`'s `parentBlock` header (validated against `fromBlock.parentHash`)
+3. Historical log accumulator at `parentBlock` based on given filters (validated with `eth_getProof`)
+4. Log accumulator at `toBlock` based on given filters (validated with `eth_getProof`)
+
+Starting from the historical log accumulator from (3), each response entry is applied to it in a way compatible with `accumulate_log` above. If the log accumulator ends up matching the value from (4), the response data is correct and `LogEntry` derived from it can be trusted.
+
+## Rationale
+
+Making the `eth_getLogs` response verifiable adds the necessary security attributes to enable wallets to transition away from relying on trusted data providers, ultimately improving the wallet's privacy guarantees as it is no longer subject to the privacy policy of any given provider.
+
+### Gas cost
+
+The gas cost produced by this scheme is significantly higher than what `LOG#` opcodes produce as of Prague, primarily due to the additional `SLOAD` / `SSTORE` requirement and the double cost of `SHA256` opcodes compared to `KECCAK256` opcodes. The gas cost increases outweigh the savings from dropping logs blooms.
```

Comment:
I don't see where any gas cost is incurred. The LOG opcode currently does not get a gas bump, the storage itself seems to be done after txs (thus a "system contract" "pays" for it), and the verification is done over RPC off-chain.

Reply:
Well, even if the system contract "pays" for it, that's extra steps that need to be done. Today's LOG opcode   similarly charges the cost for Bloom filter updates despite it being done as part of receipt / block header creation.